src/pycram/language.py

```diff
@@ -260,34 +261,50 @@ def __init__(self, condition: Union[Callable, Fluent] = None):
         """
         super().__init__(None, None)
         self.kill_event = threading.Event()
+        self.exception_queue = queue.Queue()
         if callable(condition):
             self.condition = Fluent(condition)
         elif isinstance(condition, Fluent):
             self.condition = condition
         else:
             raise AttributeError("The condition of a Monitor has to be a Callable or a Fluent")
 
-    def perform(self):
+    def perform(self) -> Tuple[State, Any]:
         """
         Behavior of the Monitor, starts a new Thread which checks the condition and then performs the attached language
         expression
 
-        :return: The result of the attached language expression
+        :return: The state of the attached language expression, as well as a list of the results of the children
         """
         def check_condition():
-            while not self.condition.get_value() and not self.kill_event.is_set():
-                time.sleep(0.1)
-            if self.kill_event.is_set():
-                return
-            for child in self.children:
-                child.interrupt()
+            while not self.kill_event.is_set():
+                try:
+                    cond = self.condition.get_value()
+                    if cond:
+                        for child in self.children:
+                            if hasattr(child, 'interrupt'):
+                                child.interrupt()
+                        if isinstance(cond, type) and issubclass(cond, Exception):
+                            self.exception_queue.put(cond)
+                        else:
+                            self.exception_queue.put(PlanFailure("Condition met in Monitor"))
+                        return
+                except Exception as e:
+                    self.exception_queue.put(e)
+                    return
+                rospy.sleep(0.1)
 
         t = threading.Thread(target=check_condition)
         t.start()
-        res = self.children[0].perform()
-        self.kill_event.set()
-        t.join()
-        return res
+        try:
+            state, result = self.children[0].perform()
+            if not self.exception_queue.empty():
+                print("Raising PlanFailure")
```

Comment:
When running the tests, this line is executed 12 times, while max_retries=6.
Is this intended? If so, please remove this print.

Reply:
done

Reply:
you are right, i removed it